### ISSUE TITLE
Resizing google.com consistently causes panic

### ISSUE THREAD

**Issue by frewsxcv** [servo/servo#8870]
Currently on ef000a458a52051b55b65e254ff9fe1a55835d04

![](https://cloud.githubusercontent.com/assets/416575/11619338/c99ab322-9c74-11e5-8c59-20ee51d39396.gif)

```
coreyf@frewbook-pro ~/D/r/servo (master)> ./mach run -r http://google.com
ERROR:js::rust: Error at https://apis.google.com/_/scs/abc-static/_/js/k=gapi.gapi.en.MWXgjJdrW-U.O/m=gapi_iframes,googleapis_client,plusone/rt=j/sv=1/d=1/ed=1/am=AAQ/rs=AItRSTMF4wQu85p8WjPpkdh4euJKa6l1Qw/cb=gapi.loaded_0:202:158: qa.Xb is not a function

thread 'PaintWorker' panicked at 'assertion failed: dst.len() >= len_src', ../src/libcore/slice.rs:1540
stack backtrace:
   1:        0x1058ec638 - sys::backtrace::tracing::imp::write::h7d2881098672eadd7It
   2:        0x1058ee77f - panicking::log_panic::_<closure>::closure.40649
   3:        0x1058ee222 - panicking::log_panic::h8037c21ca8707493MAx
   4:        0x1058d9716 - sys_common::unwind::begin_unwind_inner::h87d033c2dbe04bc77Ls
   5:        0x1058d9afe - sys_common::unwind::begin_unwind_fmt::h086b7999edf0c62fdLs
   6:        0x1058ebc57 - rust_begin_unwind
   7:        0x105910d60 - panicking::panic_fmt::h761c3d51c4df5175ZFK
   8:        0x10591105c - panicking::panic::h238a8bee891346f8wEK
   9:        0x105298e99 - _<impl>::upload::h25b9ff58c64d76a0hfa
  10:        0x10500a60c - platform::surface::_<impl>::upload::h8c8525455a971fd5Gcc
  11:        0x10480ea77 - paint_task::_<impl>::main::hda0841fd890d8fc1lkn
  12:        0x10480c2eb - sys_common::unwind::try::try_fn::try_fn::h13824183701336650925
  13:        0x1058eba78 - __rust_try
  14:        0x1058e8b7e - sys_common::unwind::try::inner_try::hdd63401dff52b04dFIs
  15:        0x10480c886 - boxed::_<impl>::call_box::call_box::h16056559659166057221
  16:        0x1058eda0d - sys::thread::_<impl>::new::thread_start::h146dc2a2095e243ajVw
  17:     0x7fff965449b0 - _pthread_body
  18:     0x7fff9654492d - _pthread_start
thread 'PaintTask PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(37) }' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', ../src/libcore/result.rs:741
stack backtrace:
   1:        0x1058ec638 - sys::backtrace::tracing::imp::write::h7d2881098672eadd7It
   2:        0x1058ee77f - panicking::log_panic::_<closure>::closure.40649
   3:        0x1058ee222 - panicking::log_panic::h8037c21ca8707493MAx
   4:        0x1058d9716 - sys_common::unwind::begin_unwind_inner::h87d033c2dbe04bc77Ls
   5:        0x1058d9afe - sys_common::unwind::begin_unwind_fmt::h086b7999edf0c62fdLs
   6:        0x1058ebc57 - rust_begin_unwind
   7:        0x105910d60 - panicking::panic_fmt::h761c3d51c4df5175ZFK
   8:        0x10481511e - paint_task::_<impl>::painted_tile_buffer::h5643e161ed697fc6jhn
   9:        0x104511767 - paint_task::_<impl>::start::start::h10784910812889635184
  10:        0x1045058be - sys_common::unwind::try::try_fn::try_fn::h7284635741652657250
  11:        0x1058eba78 - __rust_try
  12:        0x1058e8b7e - sys_common::unwind::try::inner_try::hdd63401dff52b04dFIs
  13:        0x1045063ea - boxed::_<impl>::call_box::call_box::h3389621620041994953
  14:        0x1058eda0d - sys::thread::_<impl>::new::thread_start::h146dc2a2095e243ajVw
  15:     0x7fff965449b0 - _pthread_body
  16:     0x7fff9654492d - _pthread_start
thread 'PaintWorker' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', ../src/libcore/result.rs:741
stack backtrace:
thread 'thread 'PaintWorkerPaintWorker' panicked at '' panicked at 'called `Result::unwrap()` on an `Err` value: RecvErrorcalled `Result::unwrap()` on an `Err` value: RecvError', ', ../src/libcore/result.rsthread '../src/libcore/result.rs:PaintWorker:thread '741741' panicked at 'PaintWorker
called `Result::unwrap()` on an `Err` value: RecvError
' panicked at '', called `Result::unwrap()` on an `Err` value: RecvError../src/libcore/result.rs', :../src/libcore/result.rs741:
741
thread 'Constellation' panicked at 'unable to find pipeline - this is a bug', ../src/libcore/option.rs:335
   1:        0x1058ec638 - sys::backtrace::tracing::imp::write::h7d2881098672eadd7It
   2:        0x1058ee77f - panicking::log_panic::_<closure>::closure.40649
   3:        0x1058ee222 - panicking::log_panic::h8037c21ca8707493MAx
   4:        0x1058d9716 - sys_common::unwind::begin_unwind_inner::h87d033c2dbe04bc77Ls
thread 'LayoutTask PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(37) }' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: Other, error: StringError("MPSC send failed") }) }', ../src/libcore/result.rs:741
   5:        0x1058d9afe - sys_common::unwind::begin_unwind_fmt::h086b7999edf0c62fdLs
   6:        0x1058ebc57 - rust_begin_unwind
   7:        0x105910d60 - panicking::panic_fmt::h761c3d51c4df5175ZFK
   8:        0x10480e2b8 - paint_task::_<impl>::main::hda0841fd890d8fc1lkn
   9:        0x10480c2eb - sys_common::unwind::try::try_fn::try_fn::h13824183701336650925
  10:        0x1058eba78 - __rust_try
  11:        0x1058e8b7e - sys_common::unwind::try::inner_try::hdd63401dff52b04dFIs
  12:        0x10480c886 - boxed::_<impl>::call_box::call_box::h16056559659166057221
  13:        0x1058eda0d - sys::thread::_<impl>::new::thread_start::h146dc2a2095e243ajVw
  14:     0x7fff965449b0 - _pthread_body
  15:     0x7fff9654492d - _pthread_start
stack backtrace:
   1:        0x1058ec638 - sys::backtrace::tracing::imp::write::h7d2881098672eadd7It
   2:        0x1058ee77f - panicking::log_panic::_<closure>::closure.40649
   3:        0x1058ee222 - panicking::log_panic::h8037c21ca8707493MAx
   4:        0x1058d9716 - sys_common::unwind::begin_unwind_inner::h87d033c2dbe04bc77Ls
   5:        0x1058d9afe - sys_common::unwind::begin_unwind_fmt::h086b7999edf0c62fdLs
   6:        0x1058ebc57 - rust_begin_unwind
   7:        0x105910d60 - panicking::panic_fmt::h761c3d51c4df5175ZFK
   8:        0x10480e2b8 - paint_task::_<impl>::main::hda0841fd890d8fc1lkn
   9:        0x10480c2eb - sys_common::unwind::try::try_fn::try_fn::h13824183701336650925
  10:        0x1058eba78 - __rust_try
  11:        0x1058e8b7e - sys_common::unwind::try::inner_try::hdd63401dff52b04dFIs
  12:        0x10480c886 - boxed::_<impl>::call_box::call_box::h16056559659166057221
  13:        0x1058eda0d - sys::thread::_<impl>::new::thread_start::h146dc2a2095e243ajVw
  14:     0x7fff965449b0 - _pthread_body
  15:     0x7fff9654492d - _pthread_start
stack backtrace:
   1:        0x1058ec638 - sys::backtrace::tracing::imp::write::h7d2881098672eadd7It
   2:        0x1058ee77f - panicking::log_panic::_<closure>::closure.40649
   3:        0x1058ee222 - panicking::log_panic::h8037c21ca8707493MAx
   4:        0x1058d9716 - sys_common::unwind::begin_unwind_inner::h87d033c2dbe04bc77Ls
   5:        0x1058d9afe - sys_common::unwind::begin_unwind_fmt::h086b7999edf0c62fdLs
   6:        0x1058ebc57 - rust_begin_unwind
   7:        0x105910d60 - panicking::panic_fmt::h761c3d51c4df5175ZFK
   8:        0x10480e2b8 - paint_task::_<impl>::main::hda0841fd890d8fc1lkn
   9:        0x10480c2eb - sys_common::unwind::try::try_fn::try_fn::h13824183701336650925
  10:        0x1058eba78 - __rust_try
  11:        0x1058e8b7e - sys_common::unwind::try::inner_try::hdd63401dff52b04dFIs
  12:        0x10480c886 - boxed::_<impl>::call_box::call_box::h16056559659166057221
  13:        0x1058eda0d - sys::thread::_<impl>::new::thread_start::h146dc2a2095e243ajVw
  14:     0x7fff965449b0 - _pthread_body
  15:     0x7fff9654492d - _pthread_start
stack backtrace:
   1:        0x1058ec638 - sys::backtrace::tracing::imp::write::h7d2881098672eadd7It
   2:        0x1058ee77f - panicking::log_panic::_<closure>::closure.40649
   3:        0x1058ee222 - panicking::log_panic::h8037c21ca8707493MAx
   4:        0x1058d9716 - sys_common::unwind::begin_unwind_inner::h87d033c2dbe04bc77Ls
   5:        0x1058d9afe - sys_common::unwind::begin_unwind_fmt::h086b7999edf0c62fdLs
   6:        0x1058ebc57 - rust_begin_unwind
   7:        0x105910d60 - panicking::panic_fmt::h761c3d51c4df5175ZFK
   8:        0x10480e2b8 - paint_task::_<impl>::main::hda0841fd890d8fc1lkn
   9:        0x10480c2eb - sys_common::unwind::try::try_fn::try_fn::h13824183701336650925
  10:        0x1058eba78 - __rust_try
  11:        0x1058e8b7e - sys_common::unwind::try::inner_try::hdd63401dff52b04dFIs
  12:        0x10480c886 - boxed::_<impl>::call_box::call_box::h16056559659166057221
  13:        0x1058eda0d - sys::thread::_<impl>::new::thread_start::h146dc2a2095e243ajVw
  14:     0x7fff965449b0 - _pthread_body
  15:     0x7fff9654492d - _pthread_start
stack backtrace:
   1:        0x1058ec638 - sys::backtrace::tracing::imp::write::h7d2881098672eadd7It
   2:        0x1058ee77f - panicking::log_panic::_<closure>::closure.40649
   3:        0x1058ee222 - panicking::log_panic::h8037c21ca8707493MAx
   4:        0x1058d9716 - sys_common::unwind::begin_unwind_inner::h87d033c2dbe04bc77Ls
   5:        0x1058d9afe - sys_common::unwind::begin_unwind_fmt::h086b7999edf0c62fdLs
   6:        0x1058ebc57 - rust_begin_unwind
   7:        0x105910d60 - panicking::panic_fmt::h761c3d51c4df5175ZFK
   8:        0x10480e2b8 - paint_task::_<impl>::main::hda0841fd890d8fc1lkn
   9:        0x10480c2eb - sys_common::unwind::try::try_fn::try_fn::h13824183701336650925
  10:        0x1058eba78 - __rust_try
  11:        0x1058e8b7e - sys_common::unwind::try::inner_try::hdd63401dff52b04dFIs
  12:        0x10480c886 - boxed::_<impl>::call_box::call_box::h16056559659166057221
  13:        0x1058eda0d - sys::thread::_<impl>::new::thread_start::h146dc2a2095e243ajVw
  14:     0x7fff965449b0 - _pthread_body
  15:     0x7fff9654492d - _pthread_start
stack backtrace:
   1:        0x1058ec638 - sys::backtrace::tracing::imp::write::h7d2881098672eadd7It
   2:        0x1058ee77f - panicking::log_panic::_<closure>::closure.40649
   3:        0x1058ee222 - panicking::log_panic::h8037c21ca8707493MAx
   4:        0x1058d9716 - sys_common::unwind::begin_unwind_inner::h87d033c2dbe04bc77Ls
   5:        0x1058d9afe - sys_common::unwind::begin_unwind_fmt::h086b7999edf0c62fdLs
   6:        0x1058ebc57 - rust_begin_unwind
   7:        0x105910d60 - panicking::panic_fmt::h761c3d51c4df5175ZFK
   8:        0x104364797 - constellation::_<impl>::handle_failure_msg::handle_failure_msg::h2002963372728750241
   9:        0x1042e0f84 - constellation::_<impl>::handle_request::handle_request::h17101862823283646047
  10:        0x1042d3097 - sys_common::unwind::try::try_fn::try_fn::h9926159106251436992
  11:        0x1058eba78 - __rust_try
  12:        0x1058e8b7e - sys_common::unwind::try::inner_try::hdd63401dff52b04dFIs
  13:        0x1042d515a - boxed::_<impl>::call_box::call_box::h8930793278633161271
  14:        0x1058eda0d - sys::thread::_<impl>::new::thread_start::h146dc2a2095e243ajVw
  15:     0x7fff965449b0 - _pthread_body
  16:     0x7fff9654492d - _pthread_start
stack backtrace:
   1:        0x1058ec638 - sys::backtrace::tracing::imp::write::h7d2881098672eadd7It
   2:        0x1058ee77f - panicking::log_panic::_<closure>::closure.40649
   3:        0x1058ee222 - panicking::log_panic::h8037c21ca8707493MAx
   4:        0x1058d9716 - sys_common::unwind::begin_unwind_inner::h87d033c2dbe04bc77Ls
   5:        0x1058d9afe - sys_common::unwind::begin_unwind_fmt::h086b7999edf0c62fdLs
   6:        0x1058ebc57 - rust_begin_unwind
   7:        0x105910d60 - panicking::panic_fmt::h761c3d51c4df5175ZFK
   8:        0x10472d296 - layout_task::_<impl>::exit_now::h26c44ea1d5548461VKq
   9:        0x104710f19 - layout_task::_<impl>::handle_request_helper::hc7342ac150c92ff9jAq
  10:        0x1046fabe8 - layout_task::_<impl>::handle_request::hbb22de4821e6ce80Juq
  11:        0x1046c35ca - sys_common::unwind::try::try_fn::try_fn::h11282862907011609686
  12:        0x1058eba78 - __rust_try
  13:        0x1058e8b7e - sys_common::unwind::try::inner_try::hdd63401dff52b04dFIs
  14:        0x1046c44fa - boxed::_<impl>::call_box::call_box::h14590558138400901324
  15:        0x1058eda0d - sys::thread::_<impl>::new::thread_start::h146dc2a2095e243ajVw
  16:     0x7fff965449b0 - _pthread_body
  17:     0x7fff9654492d - _pthread_start
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: "SendError(..)"', ../src/libcore/result.rs:741
stack backtrace:
   1:        0x1058ec638 - sys::backtrace::tracing::imp::write::h7d2881098672eadd7It
   2:        0x1058ee77f - panicking::log_panic::_<closure>::closure.40649
   3:        0x1058ee222 - panicking::log_panic::h8037c21ca8707493MAx
   4:        0x1058d9716 - sys_common::unwind::begin_unwind_inner::h87d033c2dbe04bc77Ls
   5:        0x1058d9afe - sys_common::unwind::begin_unwind_fmt::h086b7999edf0c62fdLs
   6:        0x1058ebc57 - rust_begin_unwind
   7:        0x105910d60 - panicking::panic_fmt::h761c3d51c4df5175ZFK
   8:        0x1042781be - compositor::_<impl>::handle_browser_message::handle_browser_message::h17818739550266500613
   9:        0x104271af7 - compositor::_<impl>::handle_events::handle_events::h4656009200612843227
  10:        0x1042a8b27 - _<impl>::handle_events::h061faf07fad60111Tda
  11:        0x10425fc65 - main::hb77cfbfb032d8741Laa
  12:        0x1058edfa2 - sys_common::unwind::try::try_fn::h7302157594252489977
  13:        0x1058eba78 - __rust_try
  14:        0x1058ede49 - rt::lang_start::h425220f787b1b804Uxx
thread 'LayoutTask PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(0) }' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError', ../src/libcore/result.rs:741
stack backtrace:
   1:        0x1058ec638 - sys::backtrace::tracing::imp::write::h7d2881098672eadd7It
Servo exited with return value 101
```

**Comment by Manishearth** [comment]
> thread 'PaintWorker' panicked at 'assertion failed: dst.len() >= len_src', ../src/libcore/slice.rs:1540

That sounded like an impossible error to me at first (There is only one error with a non-human-readable message in slice.rs; I fixed the rest -- this one isn't used in the stdlib at all), but it turns out that io-surface uses it: https://github.com/servo/io-surface-rs/blob/967224164ed27e9331fa501bab187daf1e3d49f4/src/lib.rs#L156

I guess we're messing up our threads somewhere here?

**Comment by frewsxcv** [comment]
Potentially related to #6952 and #7820

**Comment by drbo** [comment]
Could ThreadSanitizer help?

**Comment by asajeffrey** [comment]
Potentially related to #10017.

**Comment by asajeffrey** [comment]
Fixed by #10082.
